data/posts/split-route-modules.md

```diff
@@ -0,0 +1,137 @@
+---
+title: Split Route Modules
+summary: React Router v7.2's framework mode now supports automatic code splitting of route modules
+date: 2025-02-27
+image: /blog-images/headers/react-router-v7.jpg
+ogImage: /blog-images/headers/react-router-v7.jpg
+imageAlt: "The React Router logo"
+imageDisableOverlay: true
+authors:
+  - Mark Dalgleish
+---
+
+With the release of [React Router v7.2.0](https://github.com/remix-run/react-router/releases/tag/react-router%407.2.0), we’ve introduced a new opt-in framework feature called Split Route Modules. In this post we’ll explore the performance problem that Split Route Modules solves, how it works, and how to use it today.
+
+## Route Modules
+
+One of React Router’s defining features in framework mode is the [Route&nbsp;Module&nbsp;API](https://reactrouter.com/start/framework/route-module) which lets you define everything a route needs in a single file. While convenient, this API can sometimes come with a performance tradeoff.
+
+Take for example the following route module:
+
+```ts
+import { MassiveComponent } from "~/components";
+
+export async function clientLoader() {
+  return await fetch("https://example.com/api").then(
+    (response) => response.json()
+  );
+}
+
+export default function Component({ loaderData }) {
+  return <MassiveComponent data={loaderData} />;
+}
+```
+
+Here we have a minimal `clientLoader` that makes a basic fetch call to an external API. The route component, on the other hand, is much larger.
+
+Unfortunately, because these exports are both contained within the same module, the client loader has to wait for the rest of the route module to download before it can start running. If the client loader needs to perform an async operation like hitting an external API, this delay can be felt by the user.
+
+To visualize this as a timeline:
+
+```
+Get Route Module:  |--=======|
+Run clientLoader:            |-----|
+Render:                            |-|
```

Comment:
I used excalidraw for my [route.lazy](https://remix.run/blog/lazy-loading-routes#render--fetch-chains) blog post diagrams if we want to spruce these up and use something familiar from the blog.  I may have saved off the file too - let me know if you want me to look around for it.

Reply:
Yeah, I'd love to use the same diagram style as your blog post if we can.

Reply:
DM'd you a link to the excalidraw diagram

Reply:
I've updated the PR with some diagrams. Thanks for the help, it's looking much better now.